### PR TITLE
2 hotfixes wrt usedevelop and extras, 1 test setup amendment

### PR DESCRIPTION
**Pull request by hroncok**
For the love of god, I cannot run the test suite locally after upgrading to Fedora 33. Opening here to see if it works, will iterate later.

### CODE REVIEW

**Comment by hroncok** [comment]
The fixes were trivial but getting the tests to work was horrible. I also plan to add tests for the new fixes.

**Comment by hroncok** [comment]
I've added tests and verified that the following parameters of `test_self_is_not_installed` fail without the fixes:

 - `usedevelop=True`
 - `flag=--print-extras-to=-`

The other test, `test_self_is_installed_with_regular_tox` passed without the fix and is added just to ensure the implementation is not leaking.

**Comment by hroncok** [comment]
I am also testing this in Fedora CI (Zuul):

- https://src.fedoraproject.org/rpms/pyproject-rpm-macros/pull-request/129
- https://src.fedoraproject.org/rpms/python-tox-current-env/pull-request/6

**Comment by hroncok** [comment]
(rebased without changes)

File: tests/fixtures/tox.ini [hunk 1]
```diff
@@ -10,3 +10,4 @@ extras =
     full
 commands =
     python -c 'import os, sys; print(os.path.realpath(sys.exec_prefix), "is the exec_prefix")'
+# note: some tests assume [toxenv] is the last section
```
Comment (encukou) [inline tests/fixtures/tox.ini:13]:
```suggestion
# note: some tests assume [testenv] is the last section
```

**Comment by encukou** [comment]
Mostly looks good to me, except the textual manipulation of structured data. How about this?

File: tests/test_integration.py [hunk 2]
```diff
@@ -51,6 +53,24 @@ def print_extras_stdout_arg(request):
     return request.param
 
 
+@contextlib.contextmanager
+def prepare_config(tox_ini_path):
```
Comment (hroncok) [inline tests/test_integration.py:57]: Bikeshedding: What about modify_config?

File: tests/test_integration.py [hunk 3]
```diff
@@ -534,8 +563,26 @@ def test_noquiet_installed_packages(flag):
...
-        assert len([p for p in packages if p.startswith("pytest==")]) == 1
+        assert len([p for p in packages if p.startswith("tox==")]) == 1
         assert all(re.match(r"\S+==\S+", p) for p in packages)
+
+
+@pytest.mark.parametrize("flag", ["--print-deps-to=-", "--print-extras-to=-", "--current-env"])
+@pytest.mark.parametrize("usedevelop", [True, False])
+def test_self_is_not_installed(projdir, flag, usedevelop):
+    with prepare_config(projdir / 'tox.ini') as config:
+        config['testenv']['usedevelop'] = str(usedevelop)
```
Comment (hroncok) [inline tests/test_integration.py:576]:
```suggestion
        config['testenv']['usedevelop'] = usedevelop
```

File: tests/test_integration.py [hunk 4]
```diff
@@ -534,8 +563,26 @@ def test_noquiet_installed_packages(flag):
...
+    result = tox("-e", NATIVE_TOXENV, flag, quiet=False)
+    assert 'test==0.0.0' not in result.stdout
+    assert 'test @ file://' not in result.stdout
+
+
+@pytest.mark.parametrize("usedevelop", [True, False])
+def test_self_is_installed_with_regular_tox(projdir, usedevelop):
+    with prepare_config(projdir / 'tox.ini') as config:
+        config['testenv']['usedevelop'] = str(usedevelop)
```
Comment (hroncok) [inline tests/test_integration.py:585]:
```suggestion
        config['testenv']['usedevelop'] = usedevelop
```

**Review by hroncok — COMMENTED** [review]
I love the context manager. 2 slights nitpicks / suggestions.

**Comment by encukou** [comment]
Rebased with the name change, thanks for the suggestion!
ConfigParser needs strings, since INI has no standard way to encode strings. `"True"`/`"False"` work for Tox.

**Comment by hroncok** [comment]
> ConfigParser needs strings, since INI has no standard way to encode strings.

In other words, it works by accident only? Because it seem to work fro me.

```
In [1]: import configparser                                                                                                                                                                                                      

In [2]: config = configparser.ConfigParser()                                                                                                                                                                                     

In [3]: config['testenv'] = {'usedevelop': True}                                                                                                                                                                                 

In [4]: with open('ini', 'w') as f: 
   ...:     config.write(f) 
   ...:                                                                                                                                                                                                                          

In [5]: !cat ini                                                                                                                                                                                                                 
[testenv]
usedevelop = True
```

**Comment by hroncok** [comment]
(not a blocker, just curious)

**Comment by encukou** [comment]
Hm, you're right! ConfigParser doesn't check the input in *that* case. However:
```python
>>> import configparser          
>>> config = configparser.ConfigParser()
>>> config['testenv'] = {}
>>> config['testenv']['usedevelop'] = True
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.8/configparser.py", line 1258, in __setitem__
    self._parser._validate_value_types(option=key, value=value)
  File "/usr/lib64/python3.8/configparser.py", line 1185, in _validate_value_types
    raise TypeError("option values must be strings")
TypeError: option values must be strings
```

I don't think I understand all the subtlety (or inconsistencies/bugs?). But `str` works and is what we want :)

**Review by encukou — APPROVED** [review]
+1!
Normally I'd make a merge commit to keep the individual commits and show they belong together, but that's not enabled on this repo. I'll let you do the merge.

**Comment by hroncok** [comment]
I see some typos, will do a slight rebasing first.

**Comment by hroncok** [comment]
> that's not enabled on this repo

I don't like to accidentally do the github's merge commit on single commit PRs. If you find this irritating, go ahead and enable it once again.

**Comment by encukou** [comment]
Mostly I wasn't sure about your opinion. So, a merge commit would be OK here, right?

**Comment by hroncok** [comment]
I prefer linear history (and that's what I've done here), but in cases like this I understand and respect a merge commit.

**Pull request merged**